### DISCUSSION TITLE
how to cast collision avoidance constraint as a convex one to avoid getting errors

### DISCUSSION THREAD

**Discussion by Jushan-Randy-Chen** [cvxpy/cvxpy#1948] [Q&A]
Hi,

I am currently running sequential convex programming for quadrotor trajectory optimization, in which each sub problem is a quadratic program with dynamics constraint, states cosntraint, input constraint, and collision avoidance constraint (between each quadrotor). Here is a snippet of my code for a simple 2-quadrotor case:
`elif n_agents ==2:
      # x_dims = [n_states]*n_agents

      for k in range(N):
        
        objective += cvx.quad_form(s_cvx[k,:] - s_goal, Q) + cvx.quad_form(u_cvx[k,:], R) 

        constraints += [s_cvx[k+1,:] == A[k]@s_cvx[k,:] + B[k]@u_cvx[k,:] + c[k]]

        #collision avoidance:
        dist1 = s_cvx[k,0:3]-s_cvx[k,6:9]
        
        constraints+= [0<=cvx.norm(dist1,2), -cvx.norm(dist1,2) <= 0.25]
.....

However, when I run `prob.solve()`, I get this error associated with my collision avoidance constraint : 

`
DCPError: Problem does not follow DCP rules. Specifically:
The following constraints are not DCP:
0.0 <= Pnorm(var358633[0, 0:3] + -var358633[0, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[0, 0:3] + -var358633[0, 6:9], 2)
-Pnorm(var358633[0, 0:3] + -var358633[0, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[0, 0:3] + -var358633[0, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[1, 0:3] + -var358633[1, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[1, 0:3] + -var358633[1, 6:9], 2)
-Pnorm(var358633[1, 0:3] + -var358633[1, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[1, 0:3] + -var358633[1, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[2, 0:3] + -var358633[2, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[2, 0:3] + -var358633[2, 6:9], 2)
-Pnorm(var358633[2, 0:3] + -var358633[2, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[2, 0:3] + -var358633[2, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[3, 0:3] + -var358633[3, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[3, 0:3] + -var358633[3, 6:9], 2)
-Pnorm(var358633[3, 0:3] + -var358633[3, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[3, 0:3] + -var358633[3, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[4, 0:3] + -var358633[4, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[4, 0:3] + -var358633[4, 6:9], 2)
-Pnorm(var358633[4, 0:3] + -var358633[4, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[4, 0:3] + -var358633[4, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[5, 0:3] + -var358633[5, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[5, 0:3] + -var358633[5, 6:9], 2)
-Pnorm(var358633[5, 0:3] + -var358633[5, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[5, 0:3] + -var358633[5, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[6, 0:3] + -var358633[6, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[6, 0:3] + -var358633[6, 6:9], 2)
-Pnorm(var358633[6, 0:3] + -var358633[6, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[6, 0:3] + -var358633[6, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[7, 0:3] + -var358633[7, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[7, 0:3] + -var358633[7, 6:9], 2)
-Pnorm(var358633[7, 0:3] + -var358633[7, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[7, 0:3] + -var358633[7, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[8, 0:3] + -var358633[8, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[8, 0:3] + -var358633[8, 6:9], 2)
-Pnorm(var358633[8, 0:3] + -var358633[8, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[8, 0:3] + -var358633[8, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[9, 0:3] + -var358633[9, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[9, 0:3] + -var358633[9, 6:9], 2)
-Pnorm(var358633[9, 0:3] + -var358633[9, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[9, 0:3] + -var358633[9, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[10, 0:3] + -var358633[10, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[10, 0:3] + -var358633[10, 6:9], 2)
-Pnorm(var358633[10, 0:3] + -var358633[10, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[10, 0:3] + -var358633[10, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[11, 0:3] + -var358633[11, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[11, 0:3] + -var358633[11, 6:9], 2)
-Pnorm(var358633[11, 0:3] + -var358633[11, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[11, 0:3] + -var358633[11, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[12, 0:3] + -var358633[12, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[12, 0:3] + -var358633[12, 6:9], 2)
-Pnorm(var358633[12, 0:3] + -var358633[12, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[12, 0:3] + -var358633[12, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[13, 0:3] + -var358633[13, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[13, 0:3] + -var358633[13, 6:9], 2)
-Pnorm(var358633[13, 0:3] + -var358633[13, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[13, 0:3] + -var358633[13, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[14, 0:3] + -var358633[14, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[14, 0:3] + -var358633[14, 6:9], 2)
-Pnorm(var358633[14, 0:3] + -var358633[14, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[14, 0:3] + -var358633[14, 6:9], 2) <= 0.25
0.0 <= Pnorm(var358633[15, 0:3] + -var358633[15, 6:9], 2) , because the following subexpressions are not:
|--  0.0 <= Pnorm(var358633[15, 0:3] + -var358633[15, 6:9], 2)
-Pnorm(var358633[15, 0:3] + -var358633[15, 6:9], 2) <= 0.25 , because the following subexpressions are not:
|--  -Pnorm(var358633[15, 0:3] + -var358633[15, 6:9], 2) <= 0.25
...

I wonder how I should properly cast this constraint as a convex one? My goal is that the two quadrotors are always above a certain distance away from each other.

**Comment by SteveDiamond** [comment]
This constraint is not convex: ``0<=cvx.norm(dist1,2)``. This one is also not convex ``-cvx.norm(dist1,2) <= 0.25`` but I assume the minus sign is a typo. This is expected for a collision avoidance problem. You can use DCCP: https://github.com/cvxgrp/dccp

**Reply by Jushan-Randy-Chen** [reply]
Thank you! I think I fixed the error arising from the non-convexity of the collision constraint; however, the solver returns `infeasibility status` when I run it. Does that mean the problem is not infeasible based on all my constraints overall?

**Reply by SteveDiamond** [reply]
That's correct.
